data_analysis/defect_root_cause_pairs.py

```diff
@@ -0,0 +1,105 @@
+    "           AND create_date BETWEEN CAST(:start_date AS DATE) AND GETDATE() "
+    "       GROUP BY event_id HAVING COUNT(event_id) > 1 AND COUNT(event_id) < 3)"
+    "   )"
+    "SELECT "
+    "   create_date, "
+    "   event_id, "
+    "   behaviour_pattern_id, "
+    "   strand_id, "
+    "   slab_id, "
+    "   importance,"
+    "   STUFF((SELECT ',' + CAST(signal_id AS varchar(1000)) "
+    "       FROM A AS innerTable "
+    "       WHERE innerTable.event_id = p.event_id "
+    "       FOR XML PATH('')),1,1,'') "
+    "AS signal_pattern "
+    "FROM A AS p "
+    "GROUP BY p.event_id, create_date, behaviour_pattern_id, strand_id, slab_id, importance")
+
+
+def root_cause_pairs_query(date: str, signals: tuple[str, ...]) -> str:
+    # tuple[str] vs tuple[str, ...] see: https://stackoverflow.com/questions/72001132/python-typing-tuplestr-vs-tuplestr
+
+    # f"" is formatted string literal. An overview of Python String formatting: https://realpython.com/python-f-strings/
+    # f""" """ allows writing a multiline string with no escape character \ before each new line
+    root_cause_groups = f"""
+    WITH A AS (
+        SELECT A.create_date, A.event_id, B.signal_id, A.behaviour_pattern_id, A.strand_id, A.slab_id, B.importance
+        FROM main.defect_event A INNER JOIN main.defect_root_cause B 
+            on A.event_id=B.event_id 
+        WHERE A.event_id IN (
+            SELECT event_id 
+            FROM main.defect_root_cause
+            WHERE signal_id NOT IN {signals}
+                AND slab_id IS NOT null
+                AND create_date BETWEEN CAST(\'{date}' AS DATE) and GETDATE()
+            GROUP BY event_id 
+            HAVING COUNT(event_id) > 1 AND COUNT(event_id) < 3
+        )
+    )
+    SELECT p.create_date, p.event_id, p.behaviour_pattern_id, p.strand_id, p.slab_id, p.importance, STUFF(
+        (SELECT ',' + CAST(innerTable.signal_id AS varchar(1000)) FROM A AS innerTable 
+        WHERE innerTable.event_id = p.event_id FOR XML PATH('')),1,1,'') 
+    AS signal_pattern 
+    FROM A AS p 
+    GROUP BY p.event_id, p.create_date, p.behaviour_pattern_id, p.strand_id, p.slab_id, p.importance
+    """
+    return root_cause_groups
+
+
+def extract_root_cause_pairs(
+        session: Session,
+        date: str,
+        signals_to_filter: tuple[str, ...]
+) -> pd.DataFrame:
+    # tuple[str] vs tuple[str, ...] see: https://stackoverflow.com/questions/72001132/python-typing-tuplestr-vs-tuplestr
+
+    # Option 1:
+    # Solution for binding non-scalar vars as param of string queries: https://stackoverflow.com/a/56382828
+    # text_statement = create_root_cause_groups_text_clause.bindparams(bindparam('signals', expanding=True))
+
+    # Option 2:
+    query = root_cause_pairs_query(date=date, signals=signals_to_filter)
+    with session as s:
+        # result = s.execute(text_statement, params={'signals': signals_to_filter, 'start_date': date})
+        result = s.execute(query)
```

Comment:
Wäre es vllt. auch möglich einen Filter für caster einzubauen? 
Fällt mir gerade so ein !

Die caster wären '2c' oder '1'

Reply:
Ich habe nun den Filter im **extract_root_cause_pairs** Skript eingeführt. [Hier ist der Commit](https://github.com/il-ia-ni/python_sql_tutorial/pull/4/commits/27efbdff02987ed9ee817b62ed43fe94cc6c3784) dazu.

Es gibt jedoch ein ziemlich großes Problem mit den SQLAlchemy Statements in Form der f-Strings: Ein Tuple in Python darf nur dann einen Wert haben, wenn hinter ihm ein Komma steht. Solchen Tuple schafft aber SQL nicht in der "WHERE IN ('1', )"-Anweisung zu erledigen. 

Ich habe deswegen TODOs reingeschrieben, da ansonsten man doppelt einen denselben Wert reingeben muss (siehe die Zeile 128) 😱